upgrade/upgrade-scripts/rootfs-cleanup

```diff
@@ -0,0 +1,154 @@
+            continue
+
+        #
+        # Out of an abundance of caution, verify the snapshot name matches
+        # our expectations.
+        #
+        if not re.search("^execute-upgrade.\w{7}$", name):
+            raise SystemExit("unexpected snapshot name: '{}'".format(name))
+
+    filesystems = subprocess.check_output([
+        "zfs", "list", "-r", "rpool/ROOT", "-d", "1", "-Ho", "name", "-t",
+        "filesystem"
+    ]).decode("UTF-8").splitlines()
+
+    #
+    # We remove the first entry in the list, as that'll refer to the "rpool/ROOT"
+    # dataset, which is not versioned and also not eligible for deletion.
+    #
+    filesystems = filesystems[1:]
+
+    current = subprocess.check_output(["get-appliance-version"
+                                      ]).decode("UTF-8").strip()
+
+    #
+    # Our policy is to keep two delphix versions at any given time; the
+    # currently running version, and the version used prior to the
+    # current one.
+    #
+    # By sorting the list of all filesystems and snapshots on the sytem,
+    # we can easily determine the highest two versions on the system,
+    # and skip those versions (to maintain the policy described above);
+    # this is based on the assumption that the versions of an appliance
+    # will always increase (or remain constant, in the case of rollback)
+    # after an upgrade (i.e. we never upgrade to a lower version).
+    #
+    for rootfs in sorted(filesystems + snapshots, key=rootfscmp)[:-2]:
+        #
+        # In the event of a rollback, we want to be careful to not
+        # delete the currently running version, as well as any versions
+        # that may have failed to upgrade resulting in the rollback
+        # event to begin with. Thus, to account for the exceptional case
+        # of rollback, we simply skip any version that's equal to or
+        # greater than what's currently running.
+        #
+        if dpkgcmp(version(rootfs), "ge", current):
+            break
+
+        #
+        # If a rollback event has occured, it's possible that we'll be
+        # unable to delete a snapshot or filesystem; e.g. if it's still
+        # being used as the "origin" of another filesystem. Rather than
+        # adding more complicated logic to handle this situation, we
+        # simply allow the commands to throw an error.
+        #
+        # If we fail to remove a snapshot or filesystem, this shouldn't
+        # negatively impact the appliance, apart from potentially
+        # consuming some more storage than strictly necessary.
+        #
+        # Further, as the appliance continues to be upgraded to new
+        # versions over time, these error conditions will subside; it
+        # just might take a couple upgrades to allow the previously
+        # "pinned" version to be removed and cleaned up.
+        #
+        if rootfs in filesystems:
+            subprocess.run([
+                "{}/rootfs-container".format(IMAGE_PATH), "delete",
```

Comment:
How does this code ensure that snapshots are deleted before their filesystems? It isn't clear to me how the sorted function works on the union of filesystems + snapshots lists.

Reply:
The filesystem may be removed prior to the snapshots, sure, but is that a problem? When the filesystem is removed, all the snapshots under it will be removed as well.

Reply:
Ok, I thought the delete of the filesystem will fail if it had dependencies but I understand now that it recursively destroys any snapshots on the filesystem. How about the "subprocess.run" command? When it attempts to delete the non-existent snapshot later on, will the error be suppressed allowing the loop to continue or will it exit with error?

Reply:
Yea, so if we did later in the iteration, try to delete a snapshot that was already removed, we'd likely throw an error on the call to "zfs destroy".. but since I opted to not check for errors (I'm using "run" instead of "check_call"), it shouldn't cause any issues.

We could try to ensure we delete the snapshot first, or perhaps check if it exists before we do the destroy, but IMO it's not worth the complexity. Let me know if you disagree.

Reply:
I am fine with the present code. Can we document that the subprocess.run is being used in order to ignore errors, specifically this case where the order of deletion means that snapshots could already have gotten deleted?

Reply:
Yea, I can add some more comments about that case specifically.